### PR TITLE
Add template haskell support

### PR DESCRIPTION
**Pull request by Ongy**
I have added template haskell support to hdevtools.

I have also added a field in Config, because I think there should be a way to turn template haskell of when working with untrusted code.
Currently the config option is just a dummy, since it's always true.
Can you please guide me how I would add it properly.

### CODE REVIEW

File: src/CommandLoop.hs [hunk 1]
```diff
@@ -178,17 +180,25 @@ configSession state clientSend config = do
     handleGhcError :: GHC.GhcException -> GHC.Ghc String
     handleGhcError e = return $ GHC.showGhcException e ""
 
-runCommand :: IORef State -> ClientSend -> Command -> GHC.Ghc ()
-runCommand _ clientSend (CmdCheck file) = do
+runCommand :: IORef State -> ClientSend -> Bool -> Command -> GHC.Ghc ()
```
Comment (ch1bo) [inline src/CommandLoop.hs]: I'd rather pass the whole `Config` into `runCommand` than only the TH flag. This documents a bit better and aligns the signature more with `configSession`.

File: src/CommandLoop.hs [hunk 2]
```diff
@@ -178,17 +180,25 @@ configSession state clientSend config = do
...
     let noPhase = Nothing
     target <- GHC.guessTarget file noPhase
     GHC.setTargets [target]
-    let handler err = GHC.printException err >> return GHC.Failed
-    flag <- GHC.handleSourceError handler (GHC.load GHC.LoadAllTargets)
-    liftIO $ case flag of
-        GHC.Succeeded -> clientSend (ClientExit ExitSuccess)
-        GHC.Failed -> clientSend (ClientExit (ExitFailure 1))
-runCommand _ clientSend (CmdModuleFile moduleName) = do
+    graph <- GHC.depanal [] True
```
Comment (ch1bo) [inline src/CommandLoop.hs]: Do you know by any chance how expensive the dependency analysis is?

**Comment by ch1bo** [comment]
Thanks! IIRC command args are parsed quite old school with getopts into a `HdevTools` structure.

**Comment by ch1bo** [comment]
This issue is wondering me a bit. I was working the whole week without problems using `-XTemplateHaskell` and `hdevtools` (generating lenses and aeson implementations), so I am slightly confused that you need `HscInterpreted` to solve your issues.

I could reproduce your error messages and that your commits fix this issue. However, I think it's not because of your use of TemplateHaskell, but rather of using the foreign function interface. Could this be possible?

If that is the case we should rename the commandline flags and the code locations to `--no-ffi`, `requiresFFI` etc.

File: src/FindSymbol.hs [hunk 3]
```diff
@@ -48,16 +48,16 @@ findSymbol symbol files = do
...
-      let handler err = GHC.printException err >> return GHC.Failed
-      _ <- GHC.handleSourceError handler (GHC.load GHC.LoadAllTargets)
-      return ()
+       --    loadFile = do
+       --       let noPhase = Nothing
+       --       target <- GHC.guessTarget file noPhase
+       --       GHC.setTargets [target]
+       --       let handler err = GHC.printException err >> return GHC.Failed
+       --       _ <- GHC.handleSourceError handler (GHC.load GHC.LoadAllTargets)
+       --       return ()
```
Comment (ch1bo) [inline src/FindSymbol.hs]: Did you disable this on purpose?

Comment (Ongy) [inline src/FindSymbol.hs]: I disabled this code on purpose, since I factored out loading. I forgot to remove it.

**Comment by Ongy** [comment]
In the monky repository if I try to use hdevtools in `bin/Main.hs` I also get the error.
This file does not use any FFI nor does it include any module that uses FFI, but it does use TemplateHaskell. (Monky.Version).
FFI works with hdevtools from hackage as far as I can see (I can only test from hsc2hs files, which have other issues)

**Comment by Ongy** [comment]
Also I guess you are using TemplateHaskell slices defined outside your project.
The problem arises when I use TH slices defined in the package I'm editing.
Which requires the package to be build to some extend to provide the slices. In your case the slices do not have to be compiled on demand.

File: src/ClientTest.hs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import System.Directory (doesFileExist, removeFile)
+import System.Exit
+import System.IO.Error (isDoesNotExistError)
+
+import Network (sClose, Socket)
+
+import Server
+
+clientBin :: FilePath
+clientBin = "./dist/build/hdevtools/hdevtools"
```
Comment (ch1bo) [inline src/ClientTest.hs]: I support your idea of tests, although this won't work for half of our travis jobs.

**Comment by ch1bo** [comment]
Alright, makes sense. Unfortunately my comment on the `ClientTest.hs` file doesn't show up here, please have a look.

File: src/ClientTest.hs (hunk 4 above)
Comment (Ongy) [inline src/ClientTest.hs]: Sorry I was working on thit in a different branch, wasn't supposed to be in here.

**Comment by ch1bo** [comment]
Good to hear. You could run the client function in a thread to avoid looking for the executable though.

I'll merge your PR tomorrow.

**Pull request merged**